src/machine/machine_rp2040-rtc.go

```diff
@@ -0,0 +1,256 @@
+//go:build rp2040
+
+// Implementation based on code located here:
+// https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2_common/hardware_rtc/rtc.c
+
+package machine
+
+import (
+	"device/rp"
+	"errors"
+	"runtime/interrupt"
+	"runtime/volatile"
+	"unsafe"
+)
+
+type rtcType struct {
+	clkDivM1  volatile.Register32
+	setup0    volatile.Register32
+	setup1    volatile.Register32
+	ctrl      volatile.Register32
+	irqSetup0 volatile.Register32
+	irqSetup1 volatile.Register32
+	rtc1      volatile.Register32
+	rtc0      volatile.Register32
+	intR      volatile.Register32
+	intE      volatile.Register32
+	intF      volatile.Register32
+	intS      volatile.Register32
+}
+
+var RTC = (*rtcType)(unsafe.Pointer(rp.RTC))
+
+var rtcAlarmRepeats bool
+var rtcCallback func()
+
+var ErrRtcNotRunning = errors.New("RTC not running")
+var ErrRtcInvalidTime = errors.New("invalid time for RTC")
+
+func (rtc *rtcType) running() bool {
+	return rtc.ctrl.HasBits(rp.RTC_CTRL_RTC_ACTIVE)
+}
+
+func (rtc *rtcType) init() {
+	// Get clk_rtc freq and make sure it is running
+	rtcFreq := configuredFreq[clkRTC]
+	if rtcFreq == 0 {
+		panic("rtc freq is zero")
+	}
+
+	// Take rtc out of reset now that we know clk_rtc is running
+	resetBlock(rp.RESETS_RESET_RTC)
+	unresetBlockWait(rp.RESETS_RESET_RTC)
+
+	// Set up the 1 second divider.
+	// If rtc_freq is 400 then clkdiv_m1 should be 399
+	rtcFreq -= 1
+
+	// Check the freq is not too big to divide
+	if rtcFreq > rp.RTC_CLKDIV_M1_CLKDIV_M1_Msk {
+		panic("rtc freq is too big to divide")
+	}
+
+	// Write divide value
+	rtc.clkDivM1.Set(rtcFreq)
+}
+
+func (rtc *rtcType) SetTime(t RtcTime) error {
+	if !t.IsValid() {
+		return ErrRtcInvalidTime
+	}
+
+	// Disable RTC and wait while it is still active
+	rtc.ctrl.Set(0)
+	for rtc.running() {
+	}
+
+	rtc.setup0.SetBits(uint32(t.Year) << rp.RTC_SETUP_0_YEAR_Pos)
+	rtc.setup0.SetBits(uint32(t.Month) << rp.RTC_SETUP_0_MONTH_Pos)
+	rtc.setup0.SetBits(uint32(t.Day) << rp.RTC_SETUP_0_DAY_Pos)
+
+	rtc.setup1.SetBits(uint32(t.Dotw) << rp.RTC_SETUP_1_DOTW_Pos)
+	rtc.setup1.SetBits(uint32(t.Hour) << rp.RTC_SETUP_1_HOUR_Pos)
+	rtc.setup1.SetBits(uint32(t.Min) << rp.RTC_SETUP_1_MIN_Pos)
+	rtc.setup1.SetBits(uint32(t.Sec) << rp.RTC_SETUP_1_SEC_Pos)
```

Comment:
I think this is wrong: `SetBits` just sets those bits but doesn't clear it.
A better (correct and more efficient) way to do this would be by creating a new `uint32` value with the right values and then using a single `rtc.setup0.Set()` to update it.

Reply:
Great, will do!